File: study/ts/script/scene/view/siege/PopupSiegeRankCell.ts
const { ccclass, property } = cc._decorator;

import CommonRankIcon from '../../../ui/component/CommonRankIcon'
import { TextHelper } from '../../../utils/TextHelper';
import { Colors, G_UserData } from '../../../init';
import UIHelper from '../../../utils/UIHelper';
import { Path } from '../../../utils/Path';
import { handler } from '../../../utils/handler';

@ccclass
export default class PopupSiegeRankCell extends cc.Component {

    @property({
        type: cc.Node,
        visible: true
    })
    _panelBase: cc.Node = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageBG: cc.Sprite = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageBGLight: cc.Sprite = null;

    @property({
        type: CommonRankIcon,
        visible: true
    })
    _rankIcon: CommonRankIcon = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textRank: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textName: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textHurt: cc.Label = null;

    public static RANK_BG_DARK = 4
    public static TYPE_PERSON = 1	//个人
    public static TYPE_GUILD = 2	//工会

    private _rankData;
    private _type;

    onLoad() {
        var size = this._panelBase.getContentSize();
        this.node.setContentSize(size);
        this._panelBase.on(cc.Node.EventType.TOUCH_START, handler(this, this._onPanelClick));
    }

    private _createPersonNode() {
        this._textRank.string = (this._rankData.getRank());
        var hurt = this._rankData.getHurt();
        var strHurt = TextHelper.getAmountText2(hurt);
        this._textHurt.string = (strHurt);
        this._textName.string = (this._rankData.getName());
        var officerLevel = this._rankData.getOfficer_level();
        this._textName.node.color = (Colors.getOfficialColor(officerLevel));
        UIHelper.updateTextOfficialOutline(this._textName.node, officerLevel);
    }

    private _createGuildNode() {
        this._textRank.string = (this._rankData.getRank());
        var hurt = this._rankData.getHurt();
        var strHurt = TextHelper.getAmountText2(hurt);
        this._textHurt.string = (strHurt);
        this._textName.string = (this._rankData.getName());
    }

    private _setNodeBG(rank) {
        if (rank < 4) {
            var pic = Path.getComplexRankUI('img_com_ranking0' + rank);
            UIHelper.loadTexture(this._imageBG, pic);
            this._textRank.node.active = (false);
            this._rankIcon.setRank(rank);
            this._rankIcon.node.active = (true);
            this._imageBGLight.node.active = (true);
        } else if (rank % 2 == 1) {
            var pic = Path.getCommonRankUI('img_com_board_list01a');
            UIHelper.loadTexture(this._imageBG, pic);
            this._rankIcon.node.active = (false);
            this._textRank.node.active = (true);
            this._imageBGLight.node.active = (false);
        } else {
            var pic = Path.getCommonRankUI('img_com_board_list01b');
            UIHelper.loadTexture(this._imageBG, pic);
            this._rankIcon.node.active = (false);
            this._textRank.node.active = (true);
            this._imageBGLight.node.active = (false);
        }
    }

    private _onPanelClick(sender) {
        if (this._type == PopupSiegeRankCell.TYPE_PERSON) {
            var userId = this._rankData.getUser_id();
            if (userId != G_UserData.getBase().getId()) {
                G_UserData.getBase().c2sGetUserBaseInfo(userId);
            }
        }
    }

    public refreshInfo(rankData, type) {
        this._type = type;
        this._rankData = rankData;
        if (type == PopupSiegeRankCell.TYPE_PERSON) {
            this._createPersonNode();
        } else if (type == PopupSiegeRankCell.TYPE_GUILD) {
            this._createGuildNode();
        }
        var rank = this._rankData.getRank();
        this._setNodeBG(rank);
    }
}

File: study/lua/src/app/scene/view/siege/PopupSiegeRankCell.lua
local ListViewCellBase = require("app.ui.ListViewCellBase")
local PopupSiegeRankCell = class("PopupSiegeRankCell", ListViewCellBase)
local Color = require("app.utils.Color")
local Path = require("app.utils.Path")
local TextHelper = require("app.utils.TextHelper")

PopupSiegeRankCell.RANK_BG_DARK = 4

PopupSiegeRankCell.TYPE_PERSON = 1	--个人
PopupSiegeRankCell.TYPE_GUILD = 2	--工会

function PopupSiegeRankCell:ctor()
    self._rankData = nil
    self._type = nil

    self._textRank = nil    --名次
    self._textName = nil    --名字
    self._textHurt = nil    --章节
    self._imageBG = nil     --排名背景
    self._rankIcon = nil    --排名icon

	local resource = {
		file = Path.getCSB("PopupSiegeRankCell", "siege"),
		binding = {
            _panelBase = {
				events = {{event = "touch", method = "_onPanelClick"}}
			},
		}
	}
	PopupSiegeRankCell.super.ctor(self, resource)
end

function PopupSiegeRankCell:onCreate()
    local size = self._panelBase:getContentSize()
    self:setContentSize(size)
    self._panelBase:setSwallowTouches(false)
    self._rankIcon:setVisible(false)
end

function PopupSiegeRankCell:_createPersonNode()
    self._textRank:setString(self._rankData:getRank())
    local hurt = self._rankData:getHurt()
    local strHurt = TextHelper.getAmountText2(hurt)
    -- if hurt > 100000 then
    --     strHurt = math.floor(hurt/10000)..Lang.get("siege_damage_wan")
    -- end
    self._textHurt:setString(strHurt)
    self._textName:setString(self._rankData:getName())
    local officerLevel = self._rankData:getOfficer_level()
    self._textName:setColor(Color.getOfficialColor(officerLevel))
	require("yoka.utils.UIHelper").updateTextOfficialOutline(self._textName, officerLevel)
end

function PopupSiegeRankCell:_createGuildNode()
    self._textRank:setString(self._rankData:getRank())
    local hurt = self._rankData:getHurt()
    local strHurt = TextHelper.getAmountText2(hurt)
    -- if hurt > 100000 then
    --     strHurt = math.floor(hurt/10000)..Lang.get("siege_damage_wan")
    -- end
    self._textHurt:setString(strHurt)
    self._textName:setString(self._rankData:getName())
end

function PopupSiegeRankCell:_setNodeBG(rank)
    if rank < 4 then
        local pic = Path.getComplexRankUI("img_com_ranking0"..rank)
        self._imageBG:setScale9Enabled(false)
        self._imageBG:loadTexture(pic)
        self._textRank:setVisible(false)
        self._rankIcon:setRank(rank)
        self._rankIcon:setVisible(true)
        self._imageBGLight:setVisible(true)
    elseif rank % 2 == 1 then
        local pic   = Path.getCommonRankUI("img_com_board_list01a")
        self._imageBG:loadTexture(pic)
        self._imageBG:setScale9Enabled(true)
        self._imageBG:setCapInsets(cc.rect(1, 1, 1, 1))
        self._rankIcon:setVisible(false)
        self._textRank:setVisible(true)
        self._imageBGLight:setVisible(false)
    else
        local pic   = Path.getCommonRankUI("img_com_board_list01b")
        self._imageBG:setScale9Enabled(true)
        self._imageBG:setCapInsets(cc.rect(1, 1, 1, 1))
        self._imageBG:loadTexture(pic)
        self._rankIcon:setVisible(false)
        self._textRank:setVisible(true)
        self._imageBGLight:setVisible(false)
    end
end

function PopupSiegeRankCell:_onPanelClick(sender)
	local offsetX = math.abs(sender:getTouchEndPosition().x - sender:getTouchBeganPosition().x)
	local offsetY = math.abs(sender:getTouchEndPosition().y - sender:getTouchBeganPosition().y)
	if offsetX < 20 and offsetY < 20  then
        if self._type == PopupSiegeRankCell.TYPE_PERSON then
            local userId = self._rankData:getUser_id()
            if userId ~= G_UserData:getBase():getId() then
                G_UserData:getBase():c2sGetUserBaseInfo(userId)
            end
        end
	end
end


function PopupSiegeRankCell:refreshInfo(rankData, type)
    self._type = type
    self._rankData = rankData
    if type == PopupSiegeRankCell.TYPE_PERSON then 
        self:_createPersonNode()
    elseif type == PopupSiegeRankCell.TYPE_GUILD then 
        self:_createGuildNode()
    end
    local rank = self._rankData:getRank()
    self:_setNodeBG(rank)
end


return PopupSiegeRankCell